apps/emqx/src/emqx_listeners.erl

```diff
@@ -383,7 +383,24 @@ wait_listener_stopped(ListenOn) ->
             %% NOTE
             %% Tiny chance to get a connected socket here, when some other process
             %% concurrently binds to the same port.
-            gen_tcp:close(Socket)
+            gen_tcp:close(Socket),
```

Comment:
Turns out it indeed does not work the way I thought (my bad).
1. We probably need to update the NOTE above: due to the listening socket having a backlog, getting a connected socket here should be expected.
2. It should be simpler to just do `case gen_tcp:recv(Socket, 0, _CloseTimeout = 5000) of {error, closed} -> ok; {error, Reason} -> {error, Reason} end` or something similar.

Reply:
Good idea, I use gen_tcp:recv timeout to wait socket changed(closed).